RecoEcal/EgammaCoreTools/src/EcalClustersGraph.cc

```diff
@@ -0,0 +1,444 @@
+  clFeatures["cl_iz"] = clusterLocal[2];                                                              //cl_iz
+  clFeatures["cl_seed_dEta"] = deltaEta(seed_eta, cl_eta);                                            //cl_dEta
+  clFeatures["cl_seed_dPhi"] = deltaPhi(seed_phi, cl_phi);                                            //cl_dPhi
+  clFeatures["cl_seed_dEnergy"] = seed_energy - cl_energy;                                            //cl_dEnergy
+  clFeatures["cl_seed_dEt"] = (seed_energy / std::cosh(seed_eta)) - (cl_energy / std::cosh(cl_eta));  //cl_dEt
+  clFeatures["cl_nxtals"] = cluster->hitsAndFractions().size();                                       // nxtals
+  return clFeatures;
+}
+
+DeepSCInputs::FeaturesMap EcalClustersGraph::computeWindowVariables(
+    const std::vector<DeepSCInputs::FeaturesMap>& clusters) const {
+  size_t nCls = clusters.size();
+  std::map<std::string, float> min;
+  std::map<std::string, float> max;
+  std::map<std::string, float> avg;
+  for (const auto& clFeatures : clusters) {
+    for (auto const& [key, val] : clFeatures) {
+      avg[key] += (val / nCls);
+      if (val < min[key])
+        min[key] = val;
+      if (val > max[key])
+        max[key] = val;
+    }
+  }
+  DeepSCInputs::FeaturesMap windFeatures;
+  for (auto const& el : clusters.front()) {
+    windFeatures["max_" + el.first] = max[el.first];
+    windFeatures["min_" + el.first] = min[el.first];
+    windFeatures["avg_" + el.first] = avg[el.first];
+  }
+  return windFeatures;
+}
+
+std::pair<double, double> EcalClustersGraph::computeCovariances(const CaloCluster* cluster) {
+  double numeratorEtaWidth = 0;
+  double numeratorPhiWidth = 0;
+  double clEnergy = cluster->energy();
+  double denominator = clEnergy;
+  double clEta = cluster->position().eta();
+  double clPhi = cluster->position().phi();
+  std::shared_ptr<const CaloCellGeometry> this_cell;
+  EcalRecHitCollection::const_iterator rHit;
+
+  const std::vector<std::pair<DetId, float>>& detId = cluster->hitsAndFractions();
+  // Loop over recHits associated with the given SuperCluster
+  for (std::vector<std::pair<DetId, float>>::const_iterator hit = detId.begin(); hit != detId.end(); ++hit) {
+    if (PFLayer::fromCaloID(cluster->caloID()) == PFLayer::ECAL_BARREL) {
+      rHit = recHitsEB_->find((*hit).first);
+      if (rHit == recHitsEB_->end()) {
+        continue;
+      }
+    } else if (PFLayer::fromCaloID(cluster->caloID()) == PFLayer::ECAL_ENDCAP) {
+      rHit = recHitsEE_->find((*hit).first);
+      if (rHit == recHitsEE_->end()) {
+        continue;
+      }
+    }
+
+    if (PFLayer::fromCaloID(cluster->caloID()) == PFLayer::ECAL_BARREL) {
+      this_cell = ebGeom_->getGeometry(rHit->id());
+    } else if (PFLayer::fromCaloID(cluster->caloID()) == PFLayer::ECAL_ENDCAP) {
+      this_cell = eeGeom_->getGeometry(rHit->id());
+    }
+    if (this_cell == nullptr) {
```

Comment:
I agree. This part should be refactored to define `this_cell` directly in the loop and avoid the double-checking. 
Anyway, since we are looping over the ECAL PFClusters `hitsAndFractions` collection, they should always belong to either the Barrel of Endcap. 

Given the fact that this function, which is computing the shower shapes, is not currently used in the code, may I propose to fix all these three style comments in a separate PR?  We have already one in the pipeline to improve the use of `CalibratedPFCluster` pointers which is touching the same code..